storage/connectors/redis/src/main/java/feast/storage/connectors/redis/writer/RedisCustomIO.java

```diff
@@ -83,23 +84,28 @@ public Write withBatchSize(int batchSize) {
       return this;
     }
 
-    public Write withTimeout(int timeout) {
-      this.timeout = timeout;
+    public Write withFlushFrequency(Duration frequency) {
+      this.flushFrequency = frequency;
       return this;
     }
 
     @Override
     public WriteResult expand(PCollection<FeatureRow> input) {
       PCollectionTuple redisWrite =
           input
+              .apply("FixedFlushWindow", Window.<FeatureRow>into(FixedWindows.of(flushFrequency)))
               .apply(
-                  "CollectBatchBeforeWrite",
-                  Window.<FeatureRow>into(new GlobalWindows())
-                      .triggering(Repeatedly.forever(AfterPane.elementCountAtLeast(batchSize)))
-                      .discardingFiredPanes())
-              .apply("AttachSingletonKey", WithKeys.of((Void) null))
-              .apply("GroupOntoSingleton", GroupByKey.create())
+                  "AttachFeatureReferenceKey",
+                  ParDo.of(
+                      new DoFn<FeatureRow, KV<String, FeatureRow>>() {
+                        @ProcessElement
+                        public void process(ProcessContext c) {
+                          c.output(KV.of(c.element().getFeatureSet(), c.element()));
```

Comment:
What if the rows themselves are split across workers prior to grouping. Meaning the act of grouping them by this key forces them to be collocated on a worker which causes shuffling. I think the approach makes sense, but I wonder if this grouping shouldnt be done as the first step in the Beam pipeline instead of here at the back during writes. So basically as soon as we identify the key (feature set) then we can group/collocate those elements and do all other processes on them (metrics, validation, writes).

Reply:
1. True. shuffling didn't disappear completely but with null key it was much worse.
2. I like the idea with keeping groups through the whole pipeline, but I don't know how to implement it and I'm not sure it feasible at all, since it's runner's optimization responsibility.

Reply:
Ok, lets just keep this in mind. Dont think we have to optimize it now.